Null reference exception thrown in terminal logger when iterating through target outputs

### Issue Description

A `System.NullReferenceException` is thrown in terminal logger when iterating through the set of target outputs corresponding to a project which results in a cache hit (when the QuickBuild cache plugin is enabled).

### Steps to Reproduce

Assuming the cache plugin is in use by the repo, set the environment variable `EnableQuickBuildCachePlugin` to `true` and run the following command locally:
`dotnet build <project file>`

This can also be reproduced with msbuild:
`msbuild /tl <project file>`

### Expected Behavior

The expected behavior can be seen with the following command:
`dotnet build <project file> -p:EnableQuickBuildCachePlugin=false`

With the following sample outcome:
```
Restore complete (1.1s)
  X succeeded (0.8s) → c:\src\<path>\X.exe

Build succeeded in 3.2s
```

### Actual Behavior

An internal logger exception is thrown in terminal logger and surfaces the null reference exception.
```
Restore complete (1.0s)
MSBUILD : error MSB4017: The build stopped unexpectedly because of an unexpected logger failure.
Microsoft.Build.Exceptions.InternalLoggerException: The build stopped unexpectedly because of an unexpected logger failure.
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Build.Logging.TerminalLogger.TerminalLogger.TargetFinished(Object sender, TargetFinishedEventArgs e)
   at Microsoft.Build.BackEnd.Logging.EventSourceSink.RaiseEvent[TArgs](TArgs buildEvent, ArgsHandler`1 handler, ArgsHandler`1 followUpHandler)
   --- End of inner exception stack trace ---
   at Microsoft.Build.Exceptions.InternalLoggerException.Throw(Exception innerException, BuildEventArgs e, String messageResourceName, Boolean initializationException, String[] messageArgs)
   at Microsoft.Build.BackEnd.Logging.EventSourceSink.RaiseEvent[TArgs](TArgs buildEvent, ArgsHandler`1 handler, ArgsHandler`1 followUpHandler)
   at Microsoft.Build.BackEnd.Logging.EventSourceSink.Consume(BuildEventArgs buildEvent)
   at Microsoft.Build.BackEnd.Logging.EventSourceSink.RaiseEvent[TArgs](TArgs buildEvent, ArgsHandler`1 handler, ArgsHandler`1 followUpHandler)

Workload updates are available. Run `dotnet workload list` for more information.
```

### Analysis

After testing with custom msbuild bits, `TargetOutputs` corresponding to the `TargetFinishedEventArgs` is `null`, which was not previously the case for projects which resulted in a cache hit (see [here](https://github.com/dotnet/msbuild/blob/f7cbe77b09da2c0454acfa6bd9fbb4754727ffa9/src/Build/Logging/TerminalLogger/TerminalLogger.cs#L827C1-L828C1)).

This error may have resulted from changes from #11318, but it is unclear what originally caused this issue.

### Versions & Configurations

```
dotnet --version
9.0.201

msbuild --version
MSBuild version 17.14.0-preview-25105-02+915bcd94d for .NET Framework
17.14.0.10502
```
Run from Visual Studio 2022 Developer Command Prompt v17.14.0-pre.1.0

--- a/src/Build/Logging/TerminalLogger/TerminalLogger.cs
+++ b/src/Build/Logging/TerminalLogger/TerminalLogger.cs
@@ -808,15 +808,17 @@ public sealed partial class TerminalLogger : INodeLogger
         // For cache plugin projects which result in a cache hit, ensure the output path is set
         // to the item spec corresponding to the GetTargetPath target upon completion.
         var buildEventContext = e.BuildEventContext;
+        var targetOutputs = e.TargetOutputs;
         if (_restoreContext is null
             && buildEventContext is not null
+            && targetOutputs is not null
             && _hasUsedCache
             && e.TargetName == "GetTargetPath"
             && _projects.TryGetValue(new ProjectContext(buildEventContext), out TerminalProjectInfo? project))
         {
-            if (project.IsCachePluginProject)
+            if (project is not null && project.IsCachePluginProject)
             {
-                foreach (ITaskItem output in e.TargetOutputs)
+                foreach (ITaskItem output in targetOutputs)
                 {
                     project.OutputPath = output.ItemSpec.AsMemory();
                     break;

